rpc_queue.go

```diff
@@ -0,0 +1,170 @@
+package pubsub
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+var (
+	ErrQueueCancelled    = errors.New("rpc queue operation cancelled")
+	ErrQueueClosed       = errors.New("rpc queue closed")
+	ErrQueueFull         = errors.New("rpc queue full")
+	ErrQueuePushOnClosed = errors.New("push on closed rpc queue")
+)
+
+type priorityQueue struct {
+	normal   []*RPC
+	priority []*RPC
+}
+
+func (q *priorityQueue) Len() int {
+	return len(q.normal) + len(q.priority)
+}
+
+func (q *priorityQueue) NormalPush(rpc *RPC) {
+	q.normal = append(q.normal, rpc)
+}
+
+func (q *priorityQueue) PriorityPush(rpc *RPC) {
+	q.priority = append(q.priority, rpc)
+}
+
+func (q *priorityQueue) Pop() *RPC {
+	var rpc *RPC
+
+	if len(q.priority) > 0 {
+		rpc = q.priority[0]
+		q.priority = q.priority[1:]
+	} else if len(q.normal) > 0 {
+		rpc = q.normal[0]
+		q.normal = q.normal[1:]
```

Comment:
```suggestion
		rpc = q.priority[0]
		q.normal[0] = nil
		q.priority = q.priority[1:]
	} else if len(q.normal) > 0 {
		rpc = q.normal[0]
		q.normal[0] = nil
		q.normal = q.normal[1:]
```

I'd also consider using ring-buffers to avoid constantly re-allocating/moving these slices, but that can be done later after some profiling.

Reply:
I agree that we would like a more efficient implementation that doesn't stress the garbage collector, but it gets a bit more complicated with the need of avoiding memory overcommit.

Specifically, we would like the queue to shrink if there is no buildup -- a ring buffer will not have that property and use a constant amount of memory on the queue size, which can add up with a lot of peers, most of whom are not likely to have a build up.

We discussed this with @ppopth above, and I think this is fine for now and we shouldn't block merging on this -- rather make a follow up issue to write a more optimize implementation that reduces memory consumption and gc pressure.

Reply:
These are two different things:

1. Keeping the memory of the slice around is one thing and is probably ok (the queues are expected to be short).
2. Keeping the memory of the slice elements around (heap allocated) is a waste and easy to resolve.

These elements are RPC structs, so they're not small.

Reply:
What should be the hard-coded initial constant size of the memory then? or should we implement the dynamic allocation of the ring buffer ourselves?, i.e. reallocate the ring buffer if the size exceeds the capacity.

I think this optimization needs a follow up issue for more discussion and probably a new PR.

Reply:
Oh, no. I agree we shouldn't implement any of that now. I'm just saying we should implement the first suggestion in https://github.com/libp2p/go-libp2p-pubsub/pull/553#discussion_r1705907547.

Specifically, `slice[0] = nil` before re-slicing.